File: modules/aether-motion/ios/FrameAggregator.swift
import Foundation

struct FrameWindow {
  let sampleWindowMs: Double
  let frameCount: Int
  let jankCount: Int
  let jankRatio: Double?
  let averageFrameDurationMs: Double?
  let frameOverrunP95Ms: Double?
}

final class FrameAggregator {
  private let capacity: Int
  private var frameCount = 0
  private var jankCount = 0
  private var durationSumMs = 0.0
  private var durationSamples = 0
  private var overrunsMs: [Double]
  private var overrunCount = 0
  private var overrunIndex = 0
  private var windowStartMs: Double = 0

  init(capacity: Int = 64) {
    self.capacity = capacity
    self.overrunsMs = Array(repeating: 0, count: capacity)
  }

  func reset(nowMs: Double) {
    frameCount = 0
    jankCount = 0
    durationSumMs = 0
    durationSamples = 0
    overrunCount = 0
    overrunIndex = 0
    windowStartMs = nowMs
  }

  func record(durationMs: Double?, isJank: Bool, overrunMs: Double?) {
    frameCount += 1
    if isJank {
      jankCount += 1
    }
    if let durationMs, durationMs >= 0 {
      durationSumMs += durationMs
      durationSamples += 1
    }
    if let overrunMs {
      overrunsMs[overrunIndex] = overrunMs
      overrunIndex = (overrunIndex + 1) % capacity
      if overrunCount < capacity {
        overrunCount += 1
      }
    }
  }

  func snapshot(nowMs: Double) -> FrameWindow {
    let windowMs = max(0, nowMs - windowStartMs)
    let ratio = frameCount > 0 ? Double(jankCount) / Double(frameCount) : nil
    let average = durationSamples > 0 ? durationSumMs / Double(durationSamples) : nil
    return FrameWindow(
      sampleWindowMs: windowMs,
      frameCount: frameCount,
      jankCount: jankCount,
      jankRatio: ratio,
      averageFrameDurationMs: average,
      frameOverrunP95Ms: percentileMs(95)
    )
  }

  func snapshotAndReset(nowMs: Double) -> FrameWindow {
    let window = snapshot(nowMs: nowMs)
    reset(nowMs: nowMs)
    return window
  }

  func percentileMs(_ percentile: Int) -> Double? {
    guard overrunCount > 0 else { return nil }
    let values = Array(overrunsMs.prefix(overrunCount)).sorted()
    let rank = min(max(Int(ceil(Double(percentile) / 100.0 * Double(overrunCount))) - 1, 0), overrunCount - 1)
    return values[rank]
  }
}

File: modules/aether-motion/android/src/main/java/expo/modules/aethermotion/FrameAggregator.kt
package expo.modules.aethermotion

class FrameAggregator(private val capacity: Int = 64) {
  private var frameCount = 0
  private var jankCount = 0
  private var durationSumNs = 0L
  private var durationSamples = 0
  private val overrunsNs = LongArray(capacity)
  private var overrunCount = 0
  private var overrunIndex = 0
  private var windowStartElapsedMs = 0L

  fun reset(nowElapsedMs: Long) {
    frameCount = 0
    jankCount = 0
    durationSumNs = 0L
    durationSamples = 0
    overrunCount = 0
    overrunIndex = 0
    windowStartElapsedMs = nowElapsedMs
  }

  fun record(durationNs: Long?, isJank: Boolean, overrunNs: Long?) {
    frameCount += 1
    if (isJank) {
      jankCount += 1
    }
    if (durationNs != null && durationNs >= 0) {
      durationSumNs += durationNs
      durationSamples += 1
    }
    if (overrunNs != null) {
      overrunsNs[overrunIndex] = overrunNs
      overrunIndex = (overrunIndex + 1) % capacity
      if (overrunCount < capacity) {
        overrunCount += 1
      }
    }
  }

  fun snapshotAndReset(nowElapsedMs: Long): FrameWindow {
    val window = snapshot(nowElapsedMs)
    reset(nowElapsedMs)
    return window
  }

  fun snapshot(nowElapsedMs: Long): FrameWindow {
    val windowMs = (nowElapsedMs - windowStartElapsedMs).coerceAtLeast(0L)
    val jankRatio = if (frameCount > 0) jankCount.toDouble() / frameCount.toDouble() else null
    val averageMs = if (durationSamples > 0) {
      durationSumNs.toDouble() / durationSamples.toDouble() / 1_000_000.0
    } else {
      null
    }
    return FrameWindow(
      sampleWindowMs = windowMs,
      frameCount = frameCount,
      jankCount = jankCount,
      jankRatio = jankRatio,
      averageFrameDurationMs = averageMs,
      frameOverrunP95Ms = percentileMs(95),
    )
  }

  fun percentileMs(percentile: Int): Double? {
    if (overrunCount == 0) return null
    val values = LongArray(overrunCount)
    System.arraycopy(overrunsNs, 0, values, 0, overrunCount)
    values.sort()
    val rank = kotlin.math.ceil(percentile / 100.0 * overrunCount).toInt().coerceIn(1, overrunCount) - 1
    return values[rank].toDouble() / 1_000_000.0
  }

  data class FrameWindow(
    val sampleWindowMs: Long,
    val frameCount: Int,
    val jankCount: Int,
    val jankRatio: Double?,
    val averageFrameDurationMs: Double?,
    val frameOverrunP95Ms: Double?,
  )
}
